executor/analyze.go

```diff
@@ -381,3 +401,444 @@ func (e *AnalyzeColumnsExec) buildStats() (hists []*statistics.Histogram, cms []
+			fmt.Println(ctx.Err().Error())
+			*needRebuild = true
+			return
+		}
+		for _, prop := range resp.DebugGetRegionProperties.Props {
+			if prop.Name == "num_rows" {
+				var cnt uint64
+				cnt, *err = strconv.ParseUint(prop.Value, 10, 64)
+				if *err != nil {
+					*err = errors.Trace(*err)
+					return
+				}
+				newCount := atomic.AddUint64(&e.rowCount, cnt)
+				task := &AnalyzeFastTask{
+					Location:  loc,
+					LRowCount: newCount - cnt,
+					RRowCount: newCount,
+				}
+				e.tasks <- task
+			}
+		}
+	}
+}
+
+// buildSampTask return tow variable, the first bool is whether the task meeting region error
+// and need to rebuild.
+func (e *AnalyzeFastExec) buildSampTask() (bool, error) {
+	if e.wg == nil {
+		e.wg = &sync.WaitGroup{}
+	}
+
+	// Do get regions row count.
+	bo := tikv.NewBackoffer(context.Background(), 500)
+	atomic.StoreUint64(&e.rowCount, 0)
+	needRebuildForRoutine := make([]bool, e.concurrency)
+	errs := make([]error, e.concurrency)
+	e.wg.Add(e.concurrency)
+	for i := 0; i < e.concurrency; i++ {
+		go e.getSampRegionsRowCount(bo, &needRebuildForRoutine[i], &errs[i])
+	}
+
+	store, _ := e.ctx.GetStore().(tikv.Storage)
+	e.cache = store.GetRegionCache()
+	startKey, endKey := tablecodec.GetTableHandleKeyRange(e.table.Meta().ID)
+	var loc *tikv.KeyLocation
+	var err error
+	// extract all regions contain the table
+	for loc, err = e.cache.LocateKey(bo, startKey); bytes.Compare(loc.StartKey, endKey) <= 0 && err == nil; loc, err = e.cache.LocateKey(bo, append(loc.EndKey, byte(0))) {
+		if bytes.Compare(endKey, loc.EndKey) < 0 || bytes.Compare(loc.StartKey, startKey) < 0 {
+			e.scanTasks = append(e.scanTasks, loc)
+		} else {
+			e.sampLocs <- loc
+		}
+	}
+	if err != nil {
+		return false, errors.Trace(err)
+	}
+	close(e.sampLocs)
+
+	e.wg.Wait()
+	for i := 0; i < e.concurrency; i++ {
+		if errs[i] != nil {
+			return false, errors.Trace(errs[i])
+		}
+	}
+	for i := 0; i < e.concurrency; i++ {
+		if needRebuildForRoutine[i] == true {
+			return true, nil
+		}
+	}
+
+	return false, nil
+}
+
+func (e *AnalyzeFastExec) handleBatchGetResponse(kvMap *map[string][]byte, collectors []*statistics.SampleCollector) (err error) {
+	length := int32(len(*kvMap))
+	timeZone := e.ctx.GetSessionVars().Location()
+	newCursor := atomic.AddInt32(&e.sampCursor, length)
+	hasPKInfo := 0
+	if e.pkInfo != nil {
+		hasPKInfo = 1
+	}
+	hasIdxInfo := len(e.idxsInfo)
+	samplePos := newCursor - length
+	for _, value := range *kvMap {
+		if hasPKInfo > 0 {
+			collectors[0].Samples[samplePos].Ordinal = int(samplePos)
+			collectors[0].Samples[samplePos].Value, err = tablecodec.DecodeColumnValue(value, &e.pkInfo.FieldType, timeZone)
+			if err != nil {
+				return errors.Trace(err)
+			}
+		}
+		for j, colInfo := range e.colsInfo {
+			collectors[hasPKInfo+j].Samples[samplePos].Ordinal = int(samplePos)
+			collectors[hasPKInfo+j].Samples[samplePos].Value, err = tablecodec.DecodeColumnValue(value, &colInfo.FieldType, timeZone)
```

Comment:
The same with pk? There is no need to differentiate them.

Reply:
Notice the `&colInfo.FieldType`, it's different from PK.

Reply:
But `colInfo` and `pkInfo` is the same type.

Reply:
Do you mean appending the PKInfo to colsInfo? I know they are the same type. I just distinguished them to make the code clear.

Reply:
OK.